### ISSUE TITLE
Mixins

### ISSUE THREAD

**Issue by bajtos** [strongloop/loopback-boot#79]
LoopBack provides two kinds of mixins which we need to support in loopback-boot too:
 - A regular LoopBack model - mixing process copies all properties and methods to the target model.
 - A function - mixing process calls the function after the model has been attached.

**Mixin as a model - NOT IMPLEMENTED**

loopback-boot already has mechanism for defining arbitrary models, we should reuse this mechanism for defining mixins too. Benefits: developers can apply their knowledge about how to create models to create mixins too. Tools like `yo loopback:model` and Studio will support mixin definitions OOTB.

```
// common/models/timestamp-mixin.json
{
  "name": "TimeStampMixin",
  "properties": {
    "timestamp": "Date"
  }
}

// common/models/timestamp-mixin.js
module.exports = function(TimeStampMixin) {
  // setup auto-update of timestamp on save
}
```

To support this kind of mixins, we need to extend [addAllBaseModels()](https://github.com/strongloop/loopback-boot/blob/176f96e2428486c35668cceda85943229d96f184/lib/compiler.js#L187-L210) and [sortByInheritance()](https://github.com/strongloop/loopback-boot/blob/176f96e2428486c35668cceda85943229d96f184/lib/compiler.js#L220-L243) to take into account mixed-in models in addition to base models. Once that is done, model-like mixins will work both on the server and in the browserified client.

Note that this is can be implemented independently (and should be).

**Mixin as a function**

This is the part where we need to come up with a new convention.

Ultimately, I'd like to see a solution similar to what we have for models, where the compiler understands relations between model definitions and mixin definitions, and can build optimised instructions to load only those mixins that are actually used in the app.

Another thing to consider is the new component architecture as outlined [here](https://github.com/strongloop/loopback-example-component/wiki).

I am trying to come up with a small solution that can be incrementally extended in next PRs.

How about this:

 - `compiler` takes two new options: `mixinScripts` that contains a list of source files to load, `normalizeNames` that defines the algorithm for converting file names into mixin/model names.
 - mixin names are built from file names by the compiler, using `normalizeNames`
 - `executor` is extended to define mixins per instructions from the compiler

Usage:

```js
boot(app, {
  appRootDir: __dirname,
  mixins: [ './mixins/timestamp.js' ]
});
```

The next step is to implement clever auto-loading of mixins from directories, as described earlier. That way the usage would become

```
// explicit
boot(app, {
  appRootDir: __dirname,
  mixinSources: [ './mixins' ]
});

// relying on the default mixinSources value
boot(app, __dirname)
```

---

To summarise the next steps as I see them now:

 1. Support mixins defined as models
 2. Support mixins defined as functions via `mixins` option
 3. Support mixins defined as functions via `mixinSources` option

---

Random notes:

 - It seems to me that the `setup` is not handled correctly by the current mixing architecture. Imagine a `TimeStampMixin` as outlined above that adds a `setup` method to configure hooks to update the timestamp. Such `setup` method would override any `setup` method provided by the target class. This is most likely a problem of Model hooks in general as the same applies to `beforeSave`.

 - How can a Model that is being mixed in access the mixin options provided by the app? It seems to me these options are discarded when the mixin is a Model instead of a function.

This issue is a follow-up for the unfinished pull request #33, the discussion there contains useful bits of information too.

**Comment by bajtos** [comment]
> Do you want me to look into it and implement this?

Yes please, that would be great.

**Comment by clarkbw** [comment]
```json
// mixins/my-mixin.json
{ 
  "module": "my-npm-mixin-module",
  "options": {
    "key1": "description of key", 
    "key2": "description of key"
  } 
}
```

Not sure if this is still progressing but I wanted to add a request for using npm modules for mixins.  It would be nice if there was an attribute in the json file that could require a module.  This would allow community developers to publish mixin modules on npm instead of copying the JS files around.  I imagine the instructions would tell people to copy the JSON file into their mixins directory to load the mixin and then use the model config per model.

I created a prototype of the classic [TimeStamp mixin](https://github.com/clarkbw/loopback-ds-timestamp-mixin) to demonstrate how this could work, though I haven't mentioned the mixins directory yet.

**Comment by bajtos** [comment]
> Using npm modules for mixins. It would be nice if there was an attribute in the json file that could require a module. This would allow community developers to publish mixin modules on npm instead of copying the JS files around. 

Yes, the implementation must support loading mixins from `node_modules`.

**Comment by robotiko** [comment]
Hi fabien, 
I just read about those mixins that you mentioned. 
I was wondering if they are in a public repo to have a look "for inspiration"  while the pr comes to place. 
Any stats/ timeseries mixins?  sounds really interesting.

Thxs 

David

**Comment by fabien** [comment]
@bajtos I rebased/resolved any conflicts here: https://github.com/fabien/loopback-boot/tree/feature/mixin-loading-update

Next up would be having the two files, an *optional* `mixin.json` and a *mandatory* `mixin.js` file per Mixin. 

@robotiko I'd have to extract the code to be more isolated first - hopefully soon, though

**Comment by fabien** [comment]
Added loading metadata from `*.json` files. @bajtos let me know what you think - perhaps we can standardize some of the metadata keys (I've followed npm's package.json as a guideline).

Regarding loading mixins from `node_modules` - yes, this should be possible, however, I think it should be possible to load multiple mixins from a module, as well as having just a single mixin wrapped as a module. Note that it's now easy to load mixins, by setting `mixinSources`.

**Comment by bajtos** [comment]
@fabien can you please send https://github.com/fabien/loopback-boot/tree/feature/mixin-loading-update as a pull request? That way I'll see it as a task in my todo list and it's also easier to comment on a PR than on a branch.

**Comment by fabien** [comment]
@bajtos I just rebased, but apparently the commit log now shows duplicate entries. Perhaps you have some pointers on how to resolve this before I create a PR?

**Comment by bajtos** [comment]
>  I just rebased, but apparently the commit log now shows duplicate entries. Perhaps you have some pointers on how to resolve this before I create a PR?

It's hard to help without seeing the commit log itself. In the worst case you can hard-reset your feature branch to the latest master and then cherry-pick the relevant commits back.

**Comment by fabien** [comment]
@bajtos the log is still at https://github.com/fabien/loopback-boot/commits/feature/mixin-loading-update - you'll spot the duplicates immediately - shall I create a new feature branch and cherry-pick commits into it?

**Comment by Jeff-Lewis** [comment]
@fabien +10 on your mixin work. This will add great value to developers using SL. This type of of work and api extensions are not so sexy but prove to be tremendously valuable. They often are a nice surprise 'gift' when developers are just about to write some custom code (often re-inventing the wheel) and discover the existing mixins available that saves them a couple hours.

It's somewhat analogous to the value of what undserscore/lodash offer to js. Just look at the [most depended on npm modules](https://www.npmjs.com/browse/depended). #2 underscore and #3 lodash.

**Comment by bajtos** [comment]
Your branch has a merge commit https://github.com/fabien/loopback-boot/commit/25198088a886aaeb2fd46efcc5c1ce7fc0f20f16 "https://github.com/fabien/loopback-boot/commit/25198088a886aaeb2fd46efcc5c1ce7fc0f20f16" that created the mess. You must have rebased incorrectly. 

> shall I create a new feature branch and cherry-pick commits into it?

You can try `git rebase -i strongloop/master && git push -f` and if it does not fix the problem, then yes please, cherry-pick the commits. Whatever works best for you got get a clean branch that can be submitted as  pull request.

**Comment by fabien** [comment]
@bajtos thanks, that worked - I don't have the slightest idea how the rebase went wrong before.

@Jeff-Lewis thanks - mixins have been at the core of all the LB work I do; I rarely implement Model logic as plain models, nearly every bit that is reusable has been implemented by a Mixin instead.

**Comment by robotiko** [comment]
@fabien Thanks a lot for such great contribution. 
I do agree with @Jeff-Lewis , this can be really useful for reusing and sharing powerful mixins that help fine tune the platform for specific needs  that don't reach the component size/isolation. 
Some of them you already pointed, softdeletion, auditing, etc.

**Comment by bajtos** [comment]
FYI, #120 has landed on master and added support for `options.mixinDirs`. The next step is to implement `options.mixinSources` that will work similarly to `options.modelSources`, @PradnyaBaviskar will be working on this feature.

**Comment by ritch** [comment]
Cool... we need a `loopback-example-mixins` though. 

https://github.com/strongloop/loopback/issues/1349

**Comment by bajtos** [comment]
#131 added support for `mixinSources`. The last missing piece implementation-wise is strongloop/loopback-workspace#198.

The task of creating an example app is tracked by https://github.com/strongloop/loopback/issues/1349, see https://github.com/strongloop/loopback/issues/750 for documentation update.

**Comment by bajtos** [comment]
The feature is available in `loopback-boot@v2.8.0`. I am closing this issue as done.

**Comment by raymondfeng** [comment]
Do we have the documentation now?

**Comment by bajtos** [comment]
> Do we have the documentation now?

@raymondfeng  see https://github.com/strongloop/loopback/issues/750

**Comment by bajtos** [comment]
Although I am not sure if strongloop/loopback#750 is tracking the right doc task, we may need to create another task to document how to create and use custom mixins.
